packages/api/cms-api/src/dam/files/file-warning.service.ts

```diff
@@ -0,0 +1,100 @@
+import { EntityManager, FilterQuery } from "@mikro-orm/postgresql";
+import { Inject, Injectable } from "@nestjs/common";
+import { EmitWarningsServiceInterface } from "src/warnings/decorators/emit-warnings.decorator";
+
+import { WarningInput } from "../../warnings/dto/warning.input";
+import { WarningService } from "../../warnings/warning.service";
+import { DamConfig } from "../dam.config";
+import { DAM_CONFIG } from "../dam.constants";
+import { FileInterface } from "./entities/file.entity";
+import { LicenseType } from "./entities/license.embeddable";
+
+@Injectable()
+export class FileWarningService implements EmitWarningsServiceInterface<FileInterface> {
+    constructor(
+        @Inject(DAM_CONFIG) private readonly config: DamConfig,
+        private readonly entityManager: EntityManager,
+        private readonly warningService: WarningService,
+    ) {}
+
+    async emitWarningsBulk() {
+        if (!this.config.enableLicenseFeature) return; // license feature not enabled, no warnings
+
+        const soonToExpireDate = new Date();
+        soonToExpireDate.setDate(soonToExpireDate.getDate() - 30);
+
+        const filterQuery: FilterQuery<FileInterface> = [
+            {
+                license: {
+                    durationTo: {
+                        $ne: soonToExpireDate,
+                    },
+                },
+            },
+        ];
+        if (this.config.requireLicense) {
+            filterQuery.push({ license: null });
+        }
+
+        let files = [];
+        let offset = 0;
+        const limit = 50;
+        do {
+            files = await this.entityManager.getRepository<FileInterface>("DamFile").find(
+                {
+                    $or: filterQuery,
+                },
+                { limit, offset },
+            );
+
+            for (const file of files) {
+                const warnings = await this.emitWarnings(file);
+                await this.warningService.saveWarnings({
```

Comment:
why do you save the warnings inside the service for emitWarningsBulk and return them for emitWarnings?

Reply:
True I'll return it for emitWarningsBulk as well. It has to be a generator function then so I don't have to save a big array in the memory.

Reply:
changed in: 80507012840490388e2b0d81fc9fe0938094b3e2